Index: varnishd/cache_esi_fetch.c
===================================================================
--- varnish/bin/varnishd/cache_esi_fetch.c	(revision 10e4f9727fe0c04a895d3e1dc6c8b56605d53ac2)
+++ varnish/bin/varnishd/cache_esi_fetch.c	(revision e837c4bcd26218489576af899f563c66e6b0511f)
@@ -118,7 +118,10 @@
 			return(-1);
 		i = VGZ_Gunzip(vg, &dp, &dl);
-		xxxassert(i == VGZ_OK || i == VGZ_END);
-		VEP_parse(sp, dp, dl);
-		sp->obj->len += dl;
+		if (i < VGZ_OK)
+			return (-1);
+		if (dl > 0) {
+			VEP_parse(sp, dp, dl);
+			sp->obj->len += dl;
+		}
 	}
 	return (1);
@@ -271,6 +274,6 @@
 			VGZ_Obuf(sp->wrk->vgz_rx, ibuf2, sizeof ibuf2);
 			i = VGZ_Gunzip(sp->wrk->vgz_rx, &dp, &dl);
-			/* XXX: check i */
-			assert(i >= VGZ_OK);
+			if (i < VGZ_OK)
+				return (-1);
 			vef->bufp = ibuf2;
 			if (dl > 0)
